wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,304 @@
+                // return a 200 OK response with other content (for example redirection to
+                // other webpage), here we check the body length here for a light validation of
+                // the nonce value.
+                if response.status_code == 200 {
+                    let body = response.body_as_string();
+                    if body.len() < 50 {
+                        return Some(body);
+                    }
+                }
+                None
+            })
+    }
+}
+
+#[async_trait::async_trait]
+impl Authenticator for CookieAuthenticator {
+    async fn authenticate(
+        &self,
+        request: &WpNetworkRequest,
+    ) -> Result<HeaderMap, AuthenticationError> {
+        // Only attempt login if the request is to the WordPress site.
+        if let (Ok(api_base_url), Ok(request_url)) = (
+            url::Url::parse(self.api_base_url.as_str()),
+            url::Url::parse(request.url.0.as_str()),
+        ) {
+            if api_base_url.host_str() != request_url.host_str() {
+                return Err(AuthenticationError::RequestUrlDoesNotMatchSite);
+            }
+        }
+
+        self.get_rest_nonce().await.map(|nonce| {
+            let mut headers = HeaderMap::new();
+            headers.insert(
+                "X-WP-Nonce",
+                nonce
+                    .try_into()
+                    .expect("This conversion should never fail since nonce is a short string"),
+            );
+            headers
+        })
+    }
+
+    async fn re_authenticate(
+        &self,
+        request: &WpNetworkRequest,
+        previous_response: &WpNetworkResponse,
+    ) -> Result<HeaderMap, AuthenticationError> {
+        *self.nonce.write().expect("Failed to unlock nonce") = None;
+
+        return self.authenticate(request).await;
+    }
+}
+
+#[derive(Debug)]
+pub struct AuthenticatedRequestExecutor {
+    authenticator: Arc<dyn Authenticator>,
+    request_executor: Arc<dyn RequestExecutor>,
+}
+
+impl AuthenticatedRequestExecutor {
+    pub fn new(
+        authenticator: Arc<dyn Authenticator>,
+        request_executor: Arc<dyn RequestExecutor>,
+    ) -> Self {
+        Self {
+            authenticator,
+            request_executor,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl RequestExecutor for AuthenticatedRequestExecutor {
+    async fn execute(
+        &self,
+        request: Arc<WpNetworkRequest>,
+    ) -> Result<WpNetworkResponse, RequestExecutionError> {
+        let mut original = (*request).clone();
+
+        // Authenticate the initial request.
+        match self.authenticator.authenticate(&original).await {
+            Ok(headers) => {
+                original.add_headers(&headers);
+            }
+            Err(error) => {
+                // Do nothing.
+                // Any authentication error will be returned later after sending the request.
+            }
```

Comment:
> I think we should either change the authenticate function to return an Option<HeaderMap> instead of a Result or lift the error to the caller.

I have changed the return value to `Option<HeaderMap>`. Because we would always send a request to the endpoint, with or without the true authentication headers. That request will give us a full REST API response, which will be parsed into an "unauthorized" `WpApiError` instance if authentication fails.